galaxyproject/galaxy
# Assorted Data Libraries fixes

This PR tries to fix #12003 (or maybe not... see below) and addresses a backend issue for #12004 that led to #12023 as a provisional fix. Along with some other small papercuts around navigating *Data Libraries*.

## Backend Changes
- There is a specific permission to "modify" libraries, this includes changing the library name, description, synopsis, and the same for the subfolders, including some metadata about the datasets. But the API only allowed admin users to perform any modification on the library level. This has been fixed in eff762cc1ad8f8fd3187fae07e6b731ec4c571da.

## UI Changes
- User with the appropriate permissions can now see the corresponding buttons on the UI.
  ![Screenshot from 2022-01-13 14-24-01](https://user-images.githubusercontent.com/46503462/149338283-4dec5d7e-3de0-4636-8fea-e9f6dc1f689b.png)
- Delete actions (even just list deleted items) are only accessible for admins in the API so they have been hidden for any other kind of user.
- Added "Go back" links in the permissions views for libraries and folders. You can always use the browser back button, but it was kind of a papercut. Unfortunately, we can not use the breadcrumbs for folders permissions because the folder model doesn't contain the ids of all the possible parent folders (like the datasets), only the names, and direct parent id. Not sure if it's worth adding it though.

## What about #12003?
Apparently, granting "manage" permissions to other non-admin users doesn't make sense with the current API. It will require the user to have access to **all roles in the system** to actually be useful (afaik) because [trying to get a role that the user doesn't own](https://github.com/galaxyproject/galaxy/blob/320c6a36b582c62b2c24aa584044ae179176cb17/lib/galaxy/webapps/galaxy/services/libraries.py#L266) will fail.
So far, the only sensible solution I see is to leave the permission management in libraries in the hands of the admins. So apart from only showing the manage buttons for admins, in my opinion, removing the manage roles input entirely will make the UI less confusing since this feature cannot be really used by non-admin users.

![removemanage](https://user-images.githubusercontent.com/46503462/149345349-495b4c6d-340f-4872-9122-c0a246c79190.png)

Anyway, I may be wrong and maybe there is an easy fix that doesn't require exposing all the roles to the user or rewriting quite some role management code, so I leave it here for discussion :smiley: 

## TODO
- [x] Decide what to do with #12003 and do it.
  - See https://github.com/galaxyproject/galaxy/pull/13159#issuecomment-1012208427


## How to test the changes?
  - Go to Data Libraries as an admin and give different permission to other users.
  - Play around Data Libraries with those users and check that all the operations that are shown can be performed by that user.

## License
